github-auth-provider/main.go

```diff
@@ -0,0 +1,207 @@
+	}
+
+	oauthProxyOpts, err := legacyOpts.ToOptions()
+	if err != nil {
+		fmt.Printf("failed to convert legacy options to new options: %v\n", err)
+		os.Exit(1)
+	}
+
+	oauthProxyOpts.Server.BindAddress = ""
+	oauthProxyOpts.MetricsServer.BindAddress = ""
+	oauthProxyOpts.Cookie.Refresh = time.Hour
+	oauthProxyOpts.Cookie.Name = "obot_access_token"
+	oauthProxyOpts.Cookie.Secret = opts.AuthCookieSecret
+	oauthProxyOpts.Cookie.Secure = strings.HasPrefix(opts.ObotServerURL, "https://")
+	oauthProxyOpts.UpstreamServers = options.UpstreamConfig{
+		Upstreams: []options.Upstream{
+			{
+				ID:            "default",
+				URI:           "http://localhost:8080/",
+				Path:          "(.*)",
+				RewriteTarget: "$1",
+			},
+		},
+	}
+	oauthProxyOpts.RawRedirectURL = opts.ObotServerURL + "/oauth2/callback"
+	oauthProxyOpts.ReverseProxy = true
+	if opts.AuthEmailDomains != "" {
+		oauthProxyOpts.EmailDomains = strings.Split(opts.AuthEmailDomains, ",")
+	}
+
+	if err = validation.Validate(oauthProxyOpts); err != nil {
+		fmt.Printf("failed to validate options: %v\n", err)
+		os.Exit(1)
+	}
+
+	oauthProxy, err := oauth2proxy.NewOAuthProxy(oauthProxyOpts, oauth2proxy.NewValidator(oauthProxyOpts.EmailDomains, oauthProxyOpts.AuthenticatedEmailsFile))
+	if err != nil {
+		fmt.Printf("failed to create oauth2 proxy: %v\n", err)
+		os.Exit(1)
+	}
+
+	port := os.Getenv("PORT")
+	if port == "" {
+		port = "9999"
+	}
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/{$}", func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte(fmt.Sprintf("http://127.0.0.1:%s", port)))
+	})
+	mux.HandleFunc("/obot-get-state", obotGetState(oauthProxy))
+	mux.HandleFunc("/obot-get-icon-url", obotGetIconURL)
+	mux.HandleFunc("/", oauthProxy.ServeHTTP)
+
+	fmt.Printf("listening on 127.0.0.1:%s\n", port)
+	if err := http.ListenAndServe("127.0.0.1:"+port, mux); err != nil {
+		fmt.Printf("failed to listen and serve: %v\n", err)
+		os.Exit(1)
+	}
+}
+
+type SerializableRequest struct {
+	Method string              `json:"method"`
+	URL    string              `json:"url"`
+	Header map[string][]string `json:"header"`
+}
+
+type SerializableState struct {
+	ExpiresOn         *time.Time `json:"expiresOn"`
+	AccessToken       string     `json:"accessToken"`
+	PreferredUsername string     `json:"preferredUsername"`
+	User              string     `json:"user"`
+	Email             string     `json:"email"`
+}
+
+func obotGetState(p *oauth2proxy.OAuthProxy) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var sr SerializableRequest
+		if err := json.NewDecoder(r.Body).Decode(&sr); err != nil {
+			http.Error(w, fmt.Sprintf("failed to decode request body: %v", err), http.StatusBadRequest)
+			return
+		}
+
+		reqObj, err := http.NewRequest(sr.Method, sr.URL, nil)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("failed to create request object: %v", err), http.StatusBadRequest)
+			return
+		}
+
+		for k, v := range sr.Header {
+			reqObj.Header[k] = v
+		}
```

Comment:
Does something like this work? Maybe a cast is needed?
```suggestion
		reqObj.Header = sr.Header
```

Reply:
done. it works